:AllowTypeErrors

def succ: int -> int
//│ succ: int -> int


x1 = false
x2 = x1
x3 = x2
//│ x1: false
//│ x2: false
//│ x3: false

:ex
succ x3
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.15: 	succ x3
//│ ║        	^^^^^^^
//│ ╟── reference of type `false` does not match type `int`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── but it flows into reference with expected type `int`
//│ ║  l.15: 	succ x3
//│ ║        	     ^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `false`
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.8: 	x2 = x1
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.8: 	x2 = x1
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.9: 	x3 = x2
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.9: 	x3 = x2
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.15: 	succ x3
//│ ║        	     ^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╙── [info] flowing into expression of type `int`
//│ res: error | int


f1 y1 = succ y1
f2 y2 = f1 y2
f3 y3 = f2 y3
//│ f1: int -> int
//│ f2: int -> int
//│ f3: int -> int

:ex
f3 true
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.66: 	f3 true
//│ ║        	^^^^^^^
//│ ╟── reference of type `true` does not match type `int`
//│ ║  l.66: 	f3 true
//│ ║        	   ^^^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.60: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `true`
//│ ╟── [info] flowing from reference of type `true`
//│ ║  l.66: 	f3 true
//│ ║        	   ^^^^
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into variable of type `?b`
//│ ╟── [info] flowing into reference of type `?c`
//│ ║  l.60: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `?d`
//│ ╟── [info] flowing into variable of type `?e`
//│ ╟── [info] flowing into variable of type `?f`
//│ ╟── [info] flowing into reference of type `?g`
//│ ║  l.59: 	f2 y2 = f1 y2
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `?h`
//│ ╟── [info] flowing into variable of type `?i`
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.58: 	f1 y1 = succ y1
//│ ║        	             ^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╙── [info] flowing into expression of type `int`
//│ res: error | int


:ex
f3 x3
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.112: 	f3 x3
//│ ║         	^^^^^
//│ ╟── reference of type `false` does not match type `int`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── but it flows into reference with expected type `int`
//│ ║  l.112: 	f3 x3
//│ ║         	   ^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.60: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `false`
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.7: 	x1 = false
//│ ║       	     ^^^^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.8: 	x2 = x1
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.8: 	x2 = x1
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.9: 	x3 = x2
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.9: 	x3 = x2
//│ ║       	     ^^
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.112: 	f3 x3
//│ ║         	   ^^
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into variable of type `?b`
//│ ╟── [info] flowing into reference of type `?c`
//│ ║  l.60: 	f3 y3 = f2 y3
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `?d`
//│ ╟── [info] flowing into variable of type `?e`
//│ ╟── [info] flowing into variable of type `?f`
//│ ╟── [info] flowing into reference of type `?g`
//│ ║  l.59: 	f2 y2 = f1 y2
//│ ║        	           ^^
//│ ╟── [info] flowing into variable of type `?h`
//│ ╟── [info] flowing into variable of type `?i`
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.58: 	f1 y1 = succ y1
//│ ║        	             ^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╙── [info] flowing into expression of type `int`
//│ res: error | int


h1 f x = f x
h2 f x = h1 f x
h3 f x = h2 f x
//│ h1: ('a -> 'b) -> 'a -> 'b
//│ h2: (nothing -> anything) -> anything -> nothing
//│ h3: (nothing -> anything) -> anything -> nothing

:ex
h3 f3 x3
//│ res: nothing


:ex
(fun x -> succ x) false
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.271: 	(fun x -> succ x) false
//│ ║         	^^^^^^^^^^^^^^^^^^^^^^^
//│ ╟── reference of type `false` does not match type `int`
//│ ║  l.271: 	(fun x -> succ x) false
//│ ║         	                  ^^^^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── from reference:
//│ ║  l.271: 	(fun x -> succ x) false
//│ ║         	               ^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `false`
//│ ╟── [info] flowing from reference of type `false`
//│ ║  l.271: 	(fun x -> succ x) false
//│ ║         	                  ^^^^^
//│ ╟── [info] flowing into variable of type `?a`
//│ ╟── [info] flowing into variable of type `int`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.271: 	(fun x -> succ x) false
//│ ║         	               ^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╙── [info] flowing into expression of type `int`
//│ res: error | int

// :d
:ex
rec def x = add x
//│ ╔══[ERROR] Type mismatch in binding of application:
//│ ║  l.305: 	rec def x = add x
//│ ║         	            ^^^^^
//│ ╟── application of type `int -> int` does not match type `int`
//│ ╟── Note: constraint arises from reference:
//│ ║  l.305: 	rec def x = add x
//│ ║         	                ^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `int -> int`
//│ ╟── [info] flowing from application of type `int -> int`
//│ ║  l.305: 	rec def x = add x
//│ ║         	            ^^^^^
//│ ╟── [info] flowing from application of type `?a`
//│ ║  l.305: 	rec def x = add x
//│ ║         	            ^^^^^
//│ ╟── [info] flowing into expression of type `?b`
//│ ╟── [info] flowing into reference of type `int`
//│ ║  l.305: 	rec def x = add x
//│ ║         	                ^
//│ ╙── [info] flowing into expression of type `int`
//│ x: int -> int


def foo: int | string
//│ foo: int | string

:ex
succ foo
//│ ╔══[ERROR] Type mismatch in application:
//│ ║  l.333: 	succ foo
//│ ║         	^^^^^^^^
//│ ╟── type `string` does not match type `int`
//│ ║  l.329: 	def foo: int | string
//│ ║         	               ^^^^^^
//│ ╟── but it flows into reference with expected type `int`
//│ ║  l.333: 	succ foo
//│ ║         	     ^^^
//│ ╟── Note: constraint arises from type reference:
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `string`
//│ ╟── [info] flowing from type `string`
//│ ║  l.329: 	def foo: int | string
//│ ║         	               ^^^^^^
//│ ╟── [info] flowing from type `int | string`
//│ ║  l.329: 	def foo: int | string
//│ ║         	         ^^^^^^^^^^^^
//│ ╟── [info] flowing from type `int | string`
//│ ║  l.329: 	def foo: int | string
//│ ║         	         ^^^^^^^^^^^^
//│ ╟── [info] flowing from reference of type `int | string`
//│ ║  l.333: 	succ foo
//│ ║         	     ^^^
//│ ╟── [info] flowing into type `int`
//│ ║  l.3: 	def succ: int -> int
//│ ║       	          ^^^
//│ ╙── [info] flowing into expression of type `int`
//│ res: error | int



class A
class B
//│ Defined class A
//│ Defined class B

def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ty00: (A & 'a | B & 'b) -> ('a, 'b,)

def ty11: ('a & A | 'a & B) -> ('a, 'a)
//│ ty11: (A & 'a | B & 'a) -> ('a, 'a,)

ty11 = ty00
//│ (A & 'a | B & 'b) -> ('a, 'b,)
//│   <:  ty11:
//│ (A & 'a | B & 'a) -> ('a, 'a,)

// :d
:ex
ty00 = ty11
//│ (A & 'a | B & 'a) -> ('a, 'a,)
//│   <:  ty00:
//│ (A & 'a | B & 'b) -> ('a, 'b,)
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.386: 	ty00 = ty11
//│ ║         	^^^^^^^^^^^
//│ ╟── type `B & 'b` is not an instance of type 'a
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                    ^^^^^^
//│ ╟── Note: constraint arises from type variable:
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                                ^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `B & 'b`
//│ ╟── [info] flowing from type `B & 'b`
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                    ^^^^^^
//│ ╟── [info] flowing from type `B & 'b`
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	          ^^^^^^^^^^^^^^^^^
//│ ╟── [info] flowing from type `B & 'b`
//│ ║  l.376: 	def ty11: ('a & A | 'a & B) -> ('a, 'a)
//│ ║         	          ^^^^^^^^^^^^^^^^^
//│ ╟── [info] flowing from expression of type `?a`
//│ ╟── [info] flowing from type `?a0`
//│ ║  l.376: 	def ty11: ('a & A | 'a & B) -> ('a, 'a)
//│ ║         	                                ^^
//│ ╟── [info] flowing into type `'a`
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                                ^^
//│ ╙── [info] flowing into expression of type `'a`
//│ ╔══[ERROR] Type mismatch in def definition:
//│ ║  l.386: 	ty00 = ty11
//│ ║         	^^^^^^^^^^^
//│ ╟── type `A & 'a` is not an instance of type 'b
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	           ^^^^^^
//│ ╟── Note: constraint arises from type variable:
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                                    ^^
//│ ╟── ========= Additional explanations below =========
//│ ╟── [info] flowing from expression of type `A & 'a`
//│ ╟── [info] flowing from type `A & 'a`
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	           ^^^^^^
//│ ╟── [info] flowing from type `A & 'a`
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	          ^^^^^^^^^^^^^^^^^
//│ ╟── [info] flowing from type `A & 'a`
//│ ║  l.376: 	def ty11: ('a & A | 'a & B) -> ('a, 'a)
//│ ║         	          ^^^^^^^^^^^^^^^^^
//│ ╟── [info] flowing from expression of type `?a`
//│ ╟── [info] flowing from type `?a0`
//│ ║  l.376: 	def ty11: ('a & A | 'a & B) -> ('a, 'a)
//│ ║         	                                    ^^
//│ ╟── [info] flowing into type `'b`
//│ ║  l.373: 	def ty00: ('a & A | 'b & B) -> ('a, 'b)
//│ ║         	                                    ^^
//│ ╙── [info] flowing into expression of type `'b`

